text/0000-introduce-a-new-signing-mechanism.md

```diff
@@ -0,0 +1,558 @@
+
+#### Signature authentication flow
+
+As part of the `gem build` command, the gem client will open the sigstore login page using your system web browser. To sign a gem, you must choose one of the identity providers that sigstore supports.
+
+<img width="60%" alt="sigstore sign-in" src="https://user-images.githubusercontent.com/33674553/151442880-563d3eaf-e965-4d98-88c9-b8c0dc1cbaed.png">
+If you are not already signed in with your selected provider, you will need to log in.
+
+<img width="60%" alt="Sign in to GitHub" src="https://user-images.githubusercontent.com/33674553/151442855-bc68dbcd-757d-40be-b38b-88d2be61e386.png">
+
+You then need to grant sigstore permission to request your email address from your provider account. This step only occurs on your first attempt to sign a gem using this provider. Only provider-verified email addresses may be used to sign gems.
+
+<img width="60%" alt="Authorize sigstore" src="https://user-images.githubusercontent.com/33674553/151442864-51d311fd-a740-4ba5-b052-36c8fb25f6de.png">
+
+Once the authorization step is complete, you may close your browser tab. The gem signing flow completes without further input.
+
+### Signing a gem after build time
+
+Gems can be signed after build time by using the `gem signatures --sign GEMNAME`. For example, to sign gem `foo` after it has been built:
+
+```console
+$ gem signatures --sign foo
+Sign your gem by selecting an identity provider in your browser. Press Ctrl-C to abort.
+Gem `foo` signed as person@example.com
+```
+
+Before signing a gem, ensure that the email address you use to sign into your identity provider also appears in your gemspec's `email` field. The gem client will return an error otherwise.  Upon installing the gem later on, only signatures associated with emails listed in the gemspec will be considered valid.
+
+Multiple gems can be signed in a single command. For example, to sign `foo` and `bar`:
+
+```console
+$ gem signatures --sign foo bar
+Sign your gem by selecting an identity provider in your browser. Press Ctrl-C to abort.
+Gem `foo` signed as person@example.com
+Gem `bar` signed as person@example.com
+```
+
+This functionality is intended for maintainers of multiple related gems, so they do not need to go through the browser-to-CLI flow multiple times.
+
+Otherwise, using `gem signatures --sign` has the same signing behavior as `gem build`, including the browser authentication flow.
+
+### Verifying a gem at install time
+
+Gem signatures are automatically verified during installation. For example, to install and verify `foo`:
+
+```console
+$ gem install foo
+Fetching foo-1.0.0.gem
+Verified signature(s) for foo-1.0.0.gem
+Successfully installed foo-1.0.0
+[… documentation output …]
+1 gem installed
+```
+
+You may also use `gem install --verify-signatures` as in previous versions of RubyGems, but this is no longer necessary to trigger the verification process.
```

Comment:
Does `--verify-signatures` refer to the old `-P` flag? If so, would it use the `HighSecurity` policy, or remove that option in favor of a single policy for everything?

Reply:
Our proposal doesn't leverage anything from the existing gem signing mechanism. `-P` and its possible values will eventually be discontinued, in our model.

In Phase 1, where the new system is opt-in, `--verify-signatures` is closest to `MediumSecurity`:  not all gems are required to be signed, but the ones having a signature will be verified.

A gem is deemed to be signed if we can find at least one signature in Rekor using the digest of the downloaded file.  Furthermore, at least one of those signatures needs to be associated to an email address listed in the file's gemspec.

Anyone can push a signature into Rekor for a given file.  We only care about the people that also appear in the gem's own spec.

By Phase 2, we'd like to host a copy of the Rekor signature right in rubygems.org itself, alongside the gem. Less of a reliance on sigstore at installation time!